arch/risc-v/src/esp32c3/esp32c3_sha.c

```diff
@@ -0,0 +1,1666 @@
+}
+
+/****************************************************************************
+ * Name: esp32c3_sha512_finish
+ *
+ * Description:
+ *   Finishes the SHA-384 or SHA-512 operation, and writes the result to
+ *   the output buffer.
+ *
+ * Input Parameters:
+ *   ctx      - The SHA-512 context to use
+ *   output   - The SHA-512 checksum result
+ *
+ * Returned Value:
+ *   OK is returned on success.
+ *   Otherwise, a negated errno value is returned.
+ *
+ ****************************************************************************/
+
+int esp32c3_sha512_finish(struct esp32c3_sha512_context_s *ctx,
+                          unsigned char output[64])
+{
+  uint8_t used;
+  uint64_t high;
+  uint64_t low;
+  uint8_t *output_buffer = (uint8_t *)output;
+
+  used = ctx->total[0] & 0x7f;
+  ctx->buffer[used++] = 0x80;
+
+  if (used <= 112)
+    {
+      memset(ctx->buffer + used, 0, 112 - used);
+    }
+  else
+    {
+      memset(ctx->buffer + used, 0, 128 - used);
+      esp32c3_sha512_block(ctx, ctx->buffer);
+      memset(ctx->buffer, 0, 112);
+    }
+
+  high = (ctx->total[0] >> 61) | (ctx->total[1] <<  3);
+  low  = (ctx->total[0] <<  3);
+
+  PUT_UINT64_BE(high, ctx->buffer, 112);
+  PUT_UINT64_BE(low,  ctx->buffer, 120);
+
+  esp32c3_sha512_block(ctx, ctx->buffer);
+
+  PUT_UINT64_BE(ctx->state[0], output_buffer,  0);
+  PUT_UINT64_BE(ctx->state[1], output_buffer,  8);
+  PUT_UINT64_BE(ctx->state[2], output_buffer, 16);
+  PUT_UINT64_BE(ctx->state[3], output_buffer, 24);
+  PUT_UINT64_BE(ctx->state[4], output_buffer, 32);
+  PUT_UINT64_BE(ctx->state[5], output_buffer, 40);
+
+  if (ctx->mode == ESP32C3_SHA3_512)
+    {
+      PUT_UINT64_BE(ctx->state[6], output_buffer, 48);
+      PUT_UINT64_BE(ctx->state[7], output_buffer, 56);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: esp32c3_sha512_free
+ *
+ * Description:
+ *   Clears a SHA-512 context.
+ *
+ * Input Parameters:
+ *   ctx      - The SHA-512 context to clear
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+void esp32c3_sha512_free(struct esp32c3_sha512_context_s *ctx)
+{
+  if (ctx == NULL)
+    {
+      return;
+    }
+
+  memset(ctx, 0, sizeof(struct esp32c3_sha1_context_s));
+}
+
+/****************************************************************************
+ * Name: esp32c3_sha_init
+ *
+ * Description:
+ *   Initialize ESP32-C3 SHA hardware.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   OK is returned on success.
+ *   Otherwise, a negated errno value is returned.
+ *
+ ****************************************************************************/
+
+int esp32c3_sha_init(void)
+{
+  if (!g_sha_inited)
+    {
+      modifyreg32(SYSTEM_PERIP_CLK_EN1_REG, 0, SYSTEM_CRYPTO_SHA_CLK_EN);
+      modifyreg32(SYSTEM_PERIP_RST_EN1_REG, SYSTEM_CRYPTO_SHA_RST, 0);
+      g_sha_inited = true;
+    }
+
+  return OK;
```

Comment:
Improve the logic so when `g_sha_init` is in use, the return is -EBUSY.
You can add an `_err` debug message

Reply:
Thanks, I have modified it.